app/connectors/connector-verifier/src/main/java/io/syndesis/verifier/api/ComponentMetaDataAdapter.java

```diff
@@ -0,0 +1,86 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ComponentMetaDataAdapter implements MetadataAdapter {
+    private static final Logger LOG = LoggerFactory.getLogger(ComponentMetaDataAdapter.class);
+
+
+    private final Class<? extends MetaDataExtension> metaDataExtensionClass;
+
+    @SuppressWarnings("PMD.AvoidUsingVolatile")
+    private volatile MetaDataExtension metaDataExtension;
+
+
+    public ComponentMetaDataAdapter() {
+        this(MetaDataExtension.class);
+    }
+
+    public ComponentMetaDataAdapter(Class<? extends MetaDataExtension> metaDataExtensionClass) {
+        this.metaDataExtensionClass = metaDataExtensionClass;
+    }
+
+    @Override
+    public SyndesisMetadata fetch(CamelContext context, String componentId, String actionId, Map<String, Object> properties) {
+        try {
+            final MetaDataExtension extension = resolveMetaDataExtension(context, componentId, actionId);
+            final Map<String, Object> extensionOptions = CollectionsUtils.removeNullValues(properties);
+            final MetaDataExtension.MetaData metaData = extension.meta(extensionOptions)
+                .orElseThrow(() -> new IllegalArgumentException("No Metadata returned by the metadata extension"));
+
+            return adapt(context, componentId, actionId, properties, metaData);
+        } catch (@SuppressWarnings("PMD.AvoidCatchingGenericException") final Exception e) {
+            throw new IllegalStateException("Unable to fetch and process metadata", e);
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidDeeplyNestedIfStmts")
+    protected MetaDataExtension resolveMetaDataExtension(CamelContext context, String componentId, String actionId) {
+        if (metaDataExtension == null) {
+            synchronized (this) {
+                if (metaDataExtension == null) {
+                    Component component = context.getComponent(componentId, true, false);
+                    if (component == null) {
+                        LOG.error("Component {} does not exist", componentId);
+                        throw  new IllegalArgumentException(
+                            String.format("Component %s does not exists", componentId)
+                        );
+                    } else {
+                        metaDataExtension = component.getExtension(metaDataExtensionClass).orElse(null);
+                        if (metaDataExtension == null) {
+                            throw  new IllegalArgumentException(
+                                String.format("Component %s does not support meta-data extension for action %s", componentId, actionId)
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        return metaDataExtension;
+    }
+
+    protected abstract SyndesisMetadata adapt(CamelContext context, String componentId, String actionId, Map<String, Object> properties, MetaDataExtension.MetaData metadata);
```

Comment:
`postProcessMetadata`?

Reply:
it is more 'adapt' or 'convert' as it translate from `MetaDataExtension.MetaData` to `SyndesisMetadata`